### PR TITLE
Dependency management changes for annotation rules.

### LINKED ISSUE #753: Controller periodically deleting and re-creating ltm policy objects
#### Description
When deploying a route object with the virtual-server.f5.com/rewrite-target-url annotation defined, the controller will periodically delete and re-add the policy object that rewrites the host header.

**CONDITIONS** This only seems to reproduce when the virtual-server.f5.com/rewrite-target-url annotation is defined.

**IMPACT** During the time between when the object is deleted and re-created, the rewrite is not in place and causes traffic interruption.

#### OpenShift Version
v3.10

#### Controller Version
v1.6.1

#### BIG-IP Version
<Version of BIG-IP being used>

#### Diagnostic Information

**STEPS TO REPRODUCE:**

--------myroute-insecure.yml--------------
```yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  annotations:
    virtual-server.f5.com/balance: fastest-node
    virtual-server.f5.com/health: |
      [
        {
          "path":     "public.openshiftlab.com/",
          "send":     "HTTP GET /health",
          "interval": 5,
          "timeout":  10
        }
      ]
    virtual-server.f5.com/rewrite-target-url: public.openshiftlab.internal
  creationTimestamp: '2018-09-03T10:00:08Z'
  labels:
    app: openshift-lab-test1
  name: openshift-lab-test1-insecure
  namespace: default
  resourceVersion: '5734960'
  selfLink: /apis/route.openshift.io/v1/namespaces/sre-devel/routes/openshift-lab-test1
  uid: 23f00e80-af60-11e8-939e-0695e66d340a
spec:
  host: public.openshiftlab.com
  port:
    targetPort: 8080
  to:
    kind: Service
    name: httpd-ex
    weight: 100
  wildcardPolicy: None
status:
  ingress:
    - conditions:
        - lastTransitionTime: '2018-09-03T10:00:09Z'
          status: 'True'
          type: Admitted
      host: public.openshiftlab.com
      routerName: router
      wildcardPolicy: None
```
```bash
ose-master-1 $ oc new-app centos/httpd-24-centos7~https://github.com/sclorg/httpd-ex
ose-master-1 $ oc expose svc/httpd-ex
ose-master-1 $ oc create -f myroute-insecure.yml
```

Notice on the BIGIP that the host header part of the ltm policy is periodically deleted and re-added:

```
[root@bigips_1:REBOOT REQUIRED:Standalone] config # while true ; do sleep 1s ; echo -n `date`:\  ; tmsh list ltm policy /ose/openshift_insecure_routes | grep -c public.openshiftlab.internal ; done
Wed Sep 19 07:54:00 PDT 2018: 1
Wed Sep 19 07:54:01 PDT 2018: 1
Wed Sep 19 07:54:03 PDT 2018: 1
Wed Sep 19 07:54:04 PDT 2018: 1
Wed Sep 19 07:54:05 PDT 2018: 0
Wed Sep 19 07:54:07 PDT 2018: 0
Wed Sep 19 07:54:08 PDT 2018: 0
Wed Sep 19 07:54:09 PDT 2018: 0
Wed Sep 19 07:54:11 PDT 2018: 0
Wed Sep 19 07:54:12 PDT 2018: 1
Wed Sep 19 07:54:14 PDT 2018: 1
Wed Sep 19 07:54:15 PDT 2018: 1
Wed Sep 19 07:54:16 PDT 2018: 1

/var/log/audit from the above test:

Sep 19 07:54:05 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #340789-4 - object 0 - create_if { policy_rule { policy_rule_policy_name "/ose/openshift_insecure_routes" policy_rule_name "openshift_route_default_openshift-lab-test1-insecure" policy_rule_description "" policy_rule_ordinal 0 } } [Status=Command OK]
Sep 19 07:54:05 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #340789-5 - object 0 - obj_delete { policy_action { policy_action_policy_name "/ose/openshift_insecure_routes" policy_action_rule_name "openshift_route_default_openshift-lab-test1-insecure" } } [Status=Command OK]
Sep 19 07:54:05 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #340789-6 - object 0 - create_if { policy_action { policy_action_policy_name "/ose/openshift_insecure_routes" policy_action_rule_name "openshift_route_default_openshift-lab-test1-insecure" policy_action_action_id 0 policy_action_tmm_event 10 policy_action_target 19 policy_action_pool "/ose/openshift_default_httpd-ex" } } [Status=Command OK]
Sep 19 07:54:05 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #340789-7 - object 0 - obj_delete { policy_condition { policy_condition_policy_name "/ose/openshift_insecure_routes" policy_condition_rule_name "openshift_route_default_openshift-lab-test1-insecure" } } [Status=Command OK]
Sep 19 07:54:05 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #340789-8 - object 0 - create_if { policy_condition { policy_condition_policy_name "/ose/openshift_insecure_routes" policy_condition_rule_name "openshift_route_default_openshift-lab-test1-insecure" policy_condition_condition_id 0 policy_condition_tmm_event 10 policy_condition_operand 10 policy_condition_selector 6 policy_condition_condition 0 policy_condition_string_values { "public.openshiftlab.com" } } } [Status=Command OK]
Sep 19 07:54:05 bigips_1 notice icrd_child[16846]: 01420002:5: AUDIT - pid=16846 user=admin folder=/ose module=(tmos)# status=[Command OK] cmd_data=modify ltm policy /ose/openshift_insecure_routes { controls replace-all-with { forwarding } strategy /Common/first-match rules replace-all-with { openshift_route_default_openshift-lab-test1-insecure { ordinal 0 actions replace-all-with { 0 { forward request pool /ose/openshift_default_httpd-ex } } conditions replace-all-with { 0 { request equals host values { public.openshiftlab.com } http-host } } description none } openshift_route_default_httpd-ex { ordinal 1 actions replace-all-with { 0 { forward request pool /ose/openshift_default_httpd-ex } } conditions replace-all-with { 0 { request equals host values { httpd-ex-default.apps.10.145.105.61.nip.io } http-host } } description none } } references { } requires replace-all-with { http } app-service none description none draft-copy none published-copy none }

^public.openshiftlab.internal replace host OMITTED here

Sep 19 07:54:12 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #341590-4 - object 0 - create_if { policy_rule { policy_rule_policy_name "/ose/openshift_insecure_routes" policy_rule_name "openshift_route_default_openshift-lab-test1-insecure" policy_rule_description "" policy_rule_ordinal 0 } } [Status=Command OK]
Sep 19 07:54:12 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #341590-5 - object 0 - obj_delete { policy_action { policy_action_policy_name "/ose/openshift_insecure_routes" policy_action_rule_name "openshift_route_default_openshift-lab-test1-insecure" } } [Status=Command OK]
Sep 19 07:54:12 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #341590-6 - object 0 - create_if { policy_action { policy_action_policy_name "/ose/openshift_insecure_routes" policy_action_rule_name "openshift_route_default_openshift-lab-test1-insecure" policy_action_action_id 0 policy_action_tmm_event 10 policy_action_target 19 policy_action_pool "/ose/openshift_default_httpd-ex" } } [Status=Command OK]
Sep 19 07:54:12 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #341590-7 - object 0 - create_if { policy_action { policy_action_policy_name "/ose/openshift_insecure_routes" policy_action_rule_name "openshift_route_default_openshift-lab-test1-insecure" policy_action_action_id 1 policy_action_tmm_event 10 policy_action_target 4 policy_action_action 2 policy_action_value "public.openshiftlab.internal" } } [Status=Command OK]
Sep 19 07:54:12 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #341590-8 - object 0 - obj_delete { policy_condition { policy_condition_policy_name "/ose/openshift_insecure_routes" policy_condition_rule_name "openshift_route_default_openshift-lab-test1-insecure" } } [Status=Command OK]
Sep 19 07:54:12 bigips_1 notice mcpd[5634]: 01070417:5: AUDIT - client tmsh, tmsh-pid-16846, user admin - transaction #341590-9 - object 0 - create_if { policy_condition { policy_condition_policy_name "/ose/openshift_insecure_routes" policy_condition_rule_name "openshift_route_default_openshift-lab-test1-insecure" policy_condition_condition_id 0 policy_condition_tmm_event 10 policy_condition_operand 10 policy_condition_selector 6 policy_condition_condition 0 policy_condition_string_values { "public.openshiftlab.com" } } } [Status=Command OK]
Sep 19 07:54:12 bigips_1 notice icrd_child[16846]: 01420002:5: AUDIT - pid=16846 user=admin folder=/ose module=(tmos)# status=[Command OK] cmd_data=modify ltm policy /ose/openshift_insecure_routes { controls replace-all-with { forwarding } strategy /Common/first-match rules replace-all-with { openshift_route_default_openshift-lab-test1-insecure { ordinal 0 actions replace-all-with { 0 { forward request pool /ose/openshift_default_httpd-ex } 1 { request value public.openshiftlab.internal replace http-host } } conditions replace-all-with { 0 { request equals host values { public.openshiftlab.com } http-host } } description none } openshift_route_default_httpd-ex { ordinal 1 actions replace-all-with { 0 { forward request pool /ose/openshift_default_httpd-ex } } conditions replace-all-with { 0 { request equals host values { httpd-ex-default.apps.10.145.105.61.nip.io } http-host } } description none } } references { } requires replace-all-with { http } app-service none description none draft-copy none published-copy none }

^public.openshiftlab.internal replace host CREATED here


BIGIP controller logs:

2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] Overwriting existing entry for backend {ServiceName:httpd-ex ServicePort:8080 Namespace:default}
2018/09/19 14:54:04 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:04 [WARNING] Overwriting existing entry for backend {ServiceName:docker-registry ServicePort:5000 Namespace:default}
2018/09/19 14:54:04 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] Overwriting existing entry for backend {ServiceName:registry-console ServicePort:9000 Namespace:default}
2018/09/19 14:54:04 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] Overwriting existing entry for backend {ServiceName:docker-registry ServicePort:5000 Namespace:default}
2018/09/19 14:54:04 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:04 [WARNING] Overwriting existing entry for backend {ServiceName:httpd-ex ServicePort:8080 Namespace:default}
2018/09/19 14:54:04 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:04 [WARNING] Overwriting existing entry for backend {ServiceName:registry-console ServicePort:9000 Namespace:default}
2018/09/19 14:54:04 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:04 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:05 [INFO] [2018-09-19 14:54:05,697 f5_cccl.resource.resource INFO] Updating ApiPolicy: /ose/openshift_insecure_routes
2018/09/19 14:54:12 [WARNING] No profile information supplied for Route 'hawkular-metrics'
2018/09/19 14:54:12 [WARNING] No profile information supplied for Route 'logging-kibana'
2018/09/19 14:54:12 [WARNING] Overwriting existing entry for backend {ServiceName:httpd-ex ServicePort:8080 Namespace:default}
2018/09/19 14:54:12 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:12 [WARNING] Overwriting existing entry for backend {ServiceName:docker-registry ServicePort:5000 Namespace:default}
2018/09/19 14:54:12 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:12 [WARNING] Overwriting existing entry for backend {ServiceName:registry-console ServicePort:9000 Namespace:default}
2018/09/19 14:54:12 [INFO] Wrote 2 Virtual Server and 0 IApp configs
2018/09/19 14:54:12 [INFO] [2018-09-19 14:54:12,605 f5_cccl.resource.resource INFO] Updating ApiPolicy: /ose/openshift_insecure_routes
```

### PR DESCRIPTION
**Pull request by sjberman**
Problem: Annotation rules were being processed poorly and causing random updates.

Solution: Utilized the ObjectDependency logic to track changes to annotation rules, rather than blindly wiping out and recreating them every time. Also removed some dead code and updated whitelist annotations to use ObjectDependency logic as well.

Fixes #753

### CODE REVIEW

File: pkg/appmanager/resourceConfig.go [hunk 1]
```diff
@@ -2425,6 +2440,9 @@ func (rc *ResourceConfig) MergeRules(mergedRulesMap map[string]map[string]merged
 	}
 
 	rules := policy.Rules
+	if len(rules) < 2 {
+		return
+	}
```
Comment (sjberman) [inline pkg/appmanager/resourceConfig.go]: Remove this.

File: pkg/appmanager/appManager.go [hunk 2]
```diff
@@ -1107,6 +1104,51 @@ func (appMgr *Manager) syncIngresses(
 						}
 					}
 				}
+				if dep.Kind == "URL-Rewrite-Annotation" || dep.Kind == "App-Root-Annotation" {
```
Comment (ryan-talley) [inline pkg/appmanager/appManager.go]: Nitpick, but we should probably make all of these dep.Kinds into constants.

File: pkg/appmanager/resourceConfig.go [hunk 3]
```diff
@@ -496,6 +500,22 @@ func NewObjectDependencies(
 			Name:      route.Spec.Host + route.Spec.Path,
 		}
 		deps[dep]++
+		if urlRewrite, ok := route.ObjectMeta.Annotations[f5VsURLRewriteAnnotation]; ok {
+			dep = ObjectDependency{
+				Kind:      "URL-Rewrite-Annotation",
```
Comment (ryan-talley) [inline pkg/appmanager/resourceConfig.go]: ditto appManager.go 1107

File: pkg/appmanager/appManager.go (hunk 2 above)
Comment (sjberman) [reply pkg/appmanager/appManager.go]: Can do.

**Pull request merged**